junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,525 @@
+ * because it extends JUnitTestMetadata. It will know which methods to run based
+ * on the info returned by Foo$scalajs$junit$hook.metadata,
+ * it will create new test instances using Foo$scalajs$junit$hook.newInstance()
+ * and it will invoke test methods using scalajs$junit$invoke methods on the
+ * module or instance.
+ */
+class ScalaJSJUnitPlugin(val global: Global) extends NscPlugin {
+  selfPlugin =>
+
+  val name: String = "Scala.js JUnit plugin"
+
+  val components: List[NscPluginComponent] =
+    List(ScalaJSJUnitPluginComponent)
+
+  val description: String = "Makes JUnit test classes invokable in Scala.js"
+
+  object ScalaJSJUnitPluginComponent
+    extends plugins.PluginComponent with transform.Transform {
+
+    protected def newTransformer(
+        unit: global.CompilationUnit): global.Transformer = {
+      new ScalaJSJUnitPluginTransformer
+    }
+
+    val global: Global = ScalaJSJUnitPlugin.this.global
+    val phaseName: String = "junit-inject"
+    val runsAfter: List[String] = List("mixin")
+    override val runsBefore: List[String] = List("jscode")
+
+    class ScalaJSJUnitPluginTransformer extends global.Transformer {
+
+      import global._
+      import rootMirror.getRequiredClass
+
+      private val testAnnotationType =
+        getRequiredClass("org.junit.Test").toType
+
+      private val fixMethodOrderAnnotationType =
+        getRequiredClass("org.junit.FixMethodOrder").toType
+
+      private val annotationWhiteList = List(
+        testAnnotationType,
+        getRequiredClass("org.junit.Before").toType,
+        getRequiredClass("org.junit.After").toType,
+        getRequiredClass("org.junit.BeforeClass").toType,
+        getRequiredClass("org.junit.AfterClass").toType,
+        getRequiredClass("org.junit.Ignore").toType
+      )
+
+      private val scalaJSJUnitTestType =
+        getRequiredClass("org.scalajs.junit.ScalaJSJUnitTest").toType
+
+      private val jUnitClassMetadataType =
+        getRequiredClass("org.scalajs.junit.JUnitClassMetadata").toType
+
+      private val jUnitTestMetadataType =
+        getRequiredClass("org.scalajs.junit.JUnitTestMetadata").toType
+
+      private val jUnitMethodMetadataTypeTree =
+        TypeTree(getRequiredClass("org.scalajs.junit.JUnitMethodMetadata").toType)
+
+      type ScalaJSPlugin = NscPlugin {
+        /* Hack to be able to access `registerModuleExports` of ScalaJSPlugin
+         * without modifying it.
+         */
+        def PrepInteropComponent: AnyRef {
+          def genExportMember(sym: Symbol): List[Tree]
+          def registerModuleExports(sym: Symbol): Unit
+          def genNamedExport(defSym: Symbol, jsName: String, pos: Position): Unit
+        }
+      }
+
+      lazy val scalaJSPlugin = {
+        global.plugins.collectFirst {
+          case pl if pl.getClass.getName == "org.scalajs.core.compiler.ScalaJSPlugin" => pl
+        }.get.asInstanceOf[ScalaJSPlugin]
```

Comment:
Maybe have a nicer error message if the `ScalaJSPlugin` is not present. (Something like: "The Scala.js JUnit plugin only works with the Scala.js plugin enabled." rather than "Called None.get")